scamp/scamp-3.c

```diff
@@ -557,39 +557,48 @@ uint shm_ping(uint dest)
 uint shm_send_msg(uint dest, sdp_msg_t *msg) // Send msg AP
 {
     vcpu_t *vcpu = sv_vcpu + dest;
-    volatile uchar flag = 0;
-    event_t *e = event_new(proc_byte_set, (uint) &flag, 2);
-    if (e == NULL) {
-        sw_error(SW_OPT);
-        return RC_BUF;          // !! not the right RC
+
+    // Wait for the box to be idle
+    if (vcpu->mbox_ap_cmd != SHM_IDLE) {
+        volatile uchar flag = 0;
+
+        // If we can't get an event, fail
+        event_t *e = event_new(proc_byte_set, (uint) &flag, 2);
+        if (e == NULL) {
+            sw_error(SW_OPT);
+            return RC_BUF;          // !! not the right RC
+        }
+
+        // Set up a timer to wait
+        uint id = e->ID;
+        timer_schedule(e, 1000);    // !! const??
+
+        // Wait for the box to be idle, or a timeout
+        while (vcpu->mbox_ap_cmd != SHM_IDLE && flag == 0) {
+            continue;
+        }
+
+        // Return error if timed out
+        if (flag != 0) {
+            return RC_TIMEOUT;
```

Comment:
I think that you need to free event 'e' before returning to avoid a memory leak.